File: read_logs.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <map>
#include <string>

namespace fs = std::filesystem;

int count_lines(const std::string &file_path) {
  std::ifstream file(file_path);
  if (!file.is_open()) {
    return 0;
  }
  int lines = 0;
  std::string line;
  while (std::getline(file, line)) {
    lines++;
  }
  return lines;
}

void parse_line(
    const std::string &file_path, int line_number,
    std::map<std::string, std::map<std::string, int>> &daily_counts) {
  std::ifstream file(file_path);
  if (!file.is_open()) {
    return;
  }

  std::string line;
  int current_line = 0;

  while (std::getline(file, line)) {
    if (current_line == line_number) {
      std::string date = line.substr(1, 10);
      std::string level = line.substr(22, line.find(':', 22) - 22);

      if (daily_counts[date].find(level) == daily_counts[date].end()) {
        daily_counts[date][level] = 0;
      }
      daily_counts[date][level] += 1;
      return;
    }
    current_line++;
  }
}

void process_file(const std::string &file_path) {
  int num_lines = count_lines(file_path);
  std::map<std::string, std::map<std::string, int>> daily_counts;

  std::cout << "Summary for file: " << file_path << "\n";
  for (int i = 0; i < num_lines; ++i) {
    parse_line(file_path, i, daily_counts);
  }

  for (const auto &[date, counts] : daily_counts) {
    std::cout << "Date: " << date << "\n";
    for (const auto &[level, count] : counts) {
      std::cout << "  " << level << ": " << count << "\n";
    }
  }
  std::cout << "\n";
}

int main() {
  std::map<std::string, std::string> files;

  for (const auto &entry : fs::directory_iterator("logs")) {
    if (entry.is_regular_file() && entry.path().extension() == ".txt") {
      files[entry.path().filename().string()] = entry.path().string();
    }
  }
  for (const auto &[filename, file_path] : files) {
    process_file(file_path);
  }
  return 0;
}

File: read_logs.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
)

func countLines(filePath string) int {
	file, err := os.Open(filePath)
	if err != nil {
		return 0
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lines := 0
	for scanner.Scan() {
		lines++
	}
	return lines
}

func parseLine(filePath string, lineNumber int, dailyCounts map[string]map[string]int) {
	file, err := os.Open(filePath)
	if err != nil {
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	currentLine := 0

	for scanner.Scan() {
		if currentLine == lineNumber {
			line := scanner.Text()
			if len(line) < 23 {
				return
			}

			date := line[1:11]
			level := line[22 : strings.Index(line[22:], ":")+22]

			if _, exists := dailyCounts[date]; !exists {
				dailyCounts[date] = make(map[string]int)
			}
			dailyCounts[date][level]++
			return
		}
		currentLine++
	}
}

func processFile(filePath string) {
	numLines := countLines(filePath)
	dailyCounts := make(map[string]map[string]int)

	fmt.Printf("Summary for file: %s\n", filePath)
	for i := 0; i < numLines; i++ {
		parseLine(filePath, i, dailyCounts)
	}

	for date, counts := range dailyCounts {
		fmt.Printf("Date: %s\n", date)
		for level, count := range counts {
			fmt.Printf("  %s: %d\n", level, count)
		}
	}
	fmt.Println()
}

func main() {
	files := make(map[string]string)

	entries, err := ioutil.ReadDir("logs")
	if err != nil {
		return
	}

	for _, entry := range entries {
		if !entry.IsDir() && filepath.Ext(entry.Name()) == ".txt" {
			files[entry.Name()] = filepath.Join("logs", entry.Name())
		}
	}

	for _, filePath := range files {
		processFile(filePath)
	}
}
